coala_langserver/langserver.py

```diff
@@ -3,76 +3,67 @@
 from .uri import path_from_uri
 from .diagnostic import output_to_diagnostics
 
 
-class ThreadingTCPServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
-    pass
+class _StreamHandlerWrapper(socketserver.StreamRequestHandler, object):
+    '""A wrapper class that is used to construct a custom handler class.""'
 
+    delegate = None
 
-class LangserverTCPTransport(socketserver.StreamRequestHandler):
+    def setup(self):
+        super(_StreamHandlerWrapper, self).setup()
+        self.delegate = self.DELEGATE_CLASS(self.rfile, self.wfile)
 
     def handle(self):
-        s = LangServer(conn=TCPReadWriter(self.rfile, self.wfile))
-        try:
-            s.listen()
-        except Exception as e:
-            tb = traceback.format_exc()
-            log('ERROR: {} {}'.format(e, tb))
+        self.delegate.start()
 
 
-class LangServer(JSONRPC2Connection):
+class LangServer(MethodDispatcher):
     '""Language server for coala base on JSON RPC.""'
 
-    def __init__(self, conn=None):
-        super().__init__(conn=conn)
+    def __init__(self, rx, tx):
         self.root_path = None
+        self._jsonrpc_stream_reader = JsonRpcStreamReader(rx)
+        self._jsonrpc_stream_writer = JsonRpcStreamWriter(tx)
+        self._endpoint = Endpoint(self, self._jsonrpc_stream_writer.write)
+        self._dispatchers = []
+        self._shutdown = False
 
-    def handle(self, _id, request):
-        '""Handle the request from language client.""'
-        log('REQUEST: ', request)
-        resp = None
-
-        if request['method'] == 'initialize':
-            resp = self.serve_initialize(request)
-        # TODO: Support did_change and did_change_watched_files.
-        # elif request["method"] == "textDocument/didChange":
-        #     resp = self.serve_change(request)
-        # elif request["method"] == "workspace/didChangeWatchedFiles":
-        #     resp = self.serve_did_change_watched_files(request)
-        elif request['method'] == 'textDocument/didSave':
-            self.serve_did_save(request)
-
-        if resp is not None:
-            self.write_response(request['id'], resp)
-
-    def serve_initialize(self, request):
+    def start(self):
+        self._jsonrpc_stream_reader.listen(self._endpoint.consume)
+
+    def m_initialize(self, rootUri=None, rootPath=None, **kargs):
         '""Serve for the initialization request.""'
-        params = request['params']
         # Notice that the root_path could be None.
-        if 'rootUri' in params:
-            self.root_path = path_from_uri(params['rootUri'])
-        elif 'rootPath' in params:
-            self.root_path = path_from_uri(params['rootPath'])
+        if rootUri is not None:
+            self.root_path = path_from_uri(rootUri)
+        elif rootPath is not None:
+            self.root_path = path_from_uri(rootPath)
         return {
             'capabilities': {
-                'textDocumentSync': 1
-            }
+                'textDocumentSync': 1,
```

Comment:
dont modify code unless you have to.

changing lines to only improve style should only be done in commits with no functional changes, and *must* be accompanied with enabling new rules which enforce those new style constraints.